doc/adplugdb.1
.\" -*- nroff -*-
.\" This library is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU Lesser General Public
.\" License as published by the Free Software Foundation; either
.\" version 2.1 of the License, or (at your option) any later version.
.\"
.\" This library is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" Lesser General Public License for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with this library; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.TH ADPLUGDB 1 "February 25, 2003" "AdPlug database maintenance utility 1.4" "User Commands"
.SH NAME
adplugdb \- AdPlug database maintenance utility
.SH SYNOPSIS
.B adplugdb
.RI "[OPTION]... COMMAND [ARGUMENT]..."
.SH DESCRIPTION
.PP
\fBadplugdb\fP maintains database files in AdPlug database format. It
can \fBadd\fP, \fBlist\fP and \fBremove\fP records within a central
database, or \fBmerge\fP a set of databases together into one single
database.
.PP
\fBadplugdb\fP always operates on a central database file. The
location of this database file is determined by first checking if the
user has a home directory. If a home directory is present, the
database file will be located in \fB~/.adplug/adplug.db\fP. If a home
directory is not present, a database file \fBadplug.db\fP will be
looked for in the current working directory.
.PP
A system-wide database file can be used instead, by specifying the
\fB-s\fP commandline option. The system-wide database file is located
in \fB/usr/local/share/adplug/adplug.db\fP and may only be manipulated
as the superuser. An arbitrary database file might be used as well, by
specifying the \fB-d\fP commandline parameter. Only one database file
may be manipulated at a time.
.SH EXIT STATUS
\fBadplugdb\fP returns with a successful exit status (\fB0\fP on most
systems) on successful operation. An unsuccessful exit status (\fB1\fP
on most systems) is returned otherwise.
.SH COMMANDS
.PP
Commands control the main operation mode of \fBadplugdb\fP. Commands
can have a number of arguments. Only one command may be specified at a
time.
.TP
.B add
This command takes a list of filenames, separated by spaces, as
arguments. Each file is examined and a record is added to the
database if the file is supported by AdPlug. By default, the record
will be of type \fBPlain\fP, unless the \fB-t\fP commandline option is
specified (see below). The default comment entry is the specified
filename. If a record for a file is already in the database, it will
be replaced by the new record.
.TP
.B list
This command takes an optional list of filenames or keys, separated by
spaces, as arguments. Each file is examined and the corresponding
record is looked up from the database and displayed on \fBstdout\fP,
in a human-readable form. If no arguments are given, all records from
the database are displayed.
.TP
.B remove
This command takes a list of filenames or keys, separated by spaces,
as arguments. Each file is examined and the corresponding record is
removed from the database.
.TP
.B merge
This command takes a list of database filenames, separated by spaces,
as arguments. Each database file is loaded and the contents are merged
and written to the central database file. The database files are
processed in the order they are specified on the commandline. Records
from databases that were specified earlier take precedence over
records from databases that were specified later. Records from the
central database take precedence over all other records. This means
that only additional records from the other databases will be added to
the central database and if a record is found that is not already in
the central database, the version from the earliest specified database
that contains this record will be taken. In no way will records ever
be overwritten in the central database.
.SH OPTIONS
.PP
The order of the option commandline parameters is not important.
.SS "Database options:"
.TP
.B -d <file>
Specify an arbitrary file to use as the central database.
.TP
.B -s
Use the system-wide database file as the central database. This option
is only present if \fBadplugdb\fP was compiled with system-wide
database file support.
.TP
.B -t <type>
Specify a record type to be used as the type for all newly added
records. Each record needs a special type to be useful to AdPlug's
players. The commandline help, displayed using the \fB-h\fP
commandline option, presents a list of types that may be specified.
.TP
.B -c
Prompt for record comment. If this option is given, the user will be
prompted and asked for each newly added record's comment.
.TP
.B -k
Specify keys instead of filenames. If this option is given, all
command arguments that normally are filenames are expected to be
record keys instead. Each record in the database has a unique
identifying key, generated from the corresponding file's contents. To
manipulate a record entry, you either must have the exact same file
and specify its name, or you specify the record's key, using this
option. Keys are specified the same way they are displayed using the
\fBlist\fP command, as CRC16:CRC32 value in hexadecimal format.
.SS "Generic options:"
.TP
.B -q, --quiet
Be more quiet.
.TP
.B -v, --verbose
Be more verbose.
.TP
.B -h, --help
Show summary of commandline commands, arguments and options.
.TP
.B -V, --version
Show version and author information of the program.
.SH AUTHOR
Simon Peter <dn.tlp@gmx.net>


NEWS
This is a brief overview of user-visible changes in AdPlug.

Changes for version 1.6:
	- Dual OPL2 and OPL3 support.
	- This version is not binary-compatible with older versions! All
	  linked applications need to be recompiled.

Changes for version 1.5:
	- New format support:
		- XSM: eXtra Simple Music
		- DRO: DOSBox Raw OPL
		- LDS: LOUDNESS Sound System
	- Added support for .adlib files to IMF player.
	- Reverted fmopl OPL2 emulator to old LGPL'd version, so AdPlug
	  is now completely free (LGPL) again. Please read the README
	  file for details.
	- Added Ken Silverman's YM3812 emulator again. The portability
	  issues have been fixed.
	- Added a test-suite.

Changes for version 1.4:
	- Documentation changed into Texinfo format.
	- New, modularized file loading framework.
	- Now using 'libbinio' for platform-independant binary I/O.
	- Added database framework for storing music file information.
	- Uses newest fmopl OPL2 emulator version 0.60 from MAME 0.66.

Changes for version 1.3:
	- Completely new build system for DOS and Windows.
	- Centralized debug management.
	- Added a whole bunch of new developer documentation.
	- Developer documentation is now using troff formatting language.
	- Much more flexible Protracker player.
	- New or corrected IMF timer rates for the following games:
		- Duke Nukem 2
	- New format support:
		- SNG: Adlib Tracker 1.0
		- CFF: BoomTracker 4.0
		- DTM: Defy Adlib Tracker
		- DMO: Twin TrackPlayer

Changes for version 1.2:
	- Added a whole bunch of new players:
		- SAT, ROL, SNG, XAD, XMS, MAD
	- SA2 player now handles version 7 files.
	- GCC 3.x compatibility.
	- Compiles on FreeBSD (and maybe many more flavors).
	- Now using GNU autotools for UNIX builds.
	- New OPL disk writer that creates RdosPlay RAW files.
	- Reordered source directory structure (removed players directory).

Changes for version 1.1:
	- Added MKJ, DFM and BAM players.
	- Removed Ken Silverman's YM3812 emulator due to portability issues.
	- Fixed memory leak in SNG player.
	- Made AdPlug a real library.
	- Added lots of documentation.


adplugdb/adplugdb.cpp
/*
 * AdPlug - Replayer for many OPL2/OPL3 audio file formats.
 * Copyright (c) 1999 - 2002 Simon Peter <dn.tlp@gmx.net>, et al.
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * adplugdb.cpp - AdPlug database maintenance utility
 * Copyright (c) 2002 Riven the Mage <riven@ok.ru>
 * Copyright (c) 2002 Simon Peter <dn.tlp@gmx.net>
 */

#include <memory.h>
#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <binfile.h>

#include "../src/adplug.h"
#include "../src/silentopl.h"
#include "../src/database.h"
#include "getopt.h"

/***** Defines *****/

#define UNKNOWN_FILETYPE	"*** Unknown ***"

// Message urgency levels
#define MSG_PANIC	0
#define MSG_ERROR	1
#define MSG_WARN	2
#define MSG_NOTE	3
#define MSG_DEBUG	4

/***** Global variables *****/

static const struct {
  const char				*typestr;
  CAdPlugDatabase::CRecord::RecordType	type;
} rtypes[] = {
  { "plain", CAdPlugDatabase::CRecord::Plain },
  { "songinfo", CAdPlugDatabase::CRecord::SongInfo },
  { "clockspeed", CAdPlugDatabase::CRecord::ClockSpeed },
  {0}
};

static struct {
  const char *db_file;
  CAdPlugDatabase::CRecord::RecordType rtype;
  int message_level;
} cfg = {
  "adplug.db",
  CAdPlugDatabase::CRecord::Plain,
  MSG_NOTE
};

static CAdPlugDatabase	mydb;
static const char	*program_name;

/***** Functions *****/

static void message(int level, const char *fmt, ...)
{
  va_list argptr;

  if(cfg.message_level < level) return;

  fprintf(stderr, "%s: ", program_name);
  va_start(argptr, fmt);
  vfprintf(stderr, fmt, argptr);
  va_end(argptr);
  fprintf(stderr, "\n");
}

static CAdPlugDatabase::CKey file2key(const char *filename)
{
  binifstream f(filename);

  if(f.error()) {
    message(MSG_ERROR, "can't open specified file -- %s", filename);
    exit(EXIT_FAILURE);
  }

  return CAdPlugDatabase::CKey(f);
}

static void usage()
{
  printf("Usage: %s [options] <command> [arguments]\n\n"
	 "Commands:\n"
	 "  add <files>      Add (and replace) files to database\n"
	 "  list [files]     List files (or everything) from database\n"
	 "  remove <files>   Remove files from database\n"
	 "\n"
	 "Database options:\n"
	 "  -d <file>        Use different database file\n"
	 "  -t <type>        Add as different record type\n"
	 "\n"
	 "Generic options:\n"
	 "  -q               Be more quiet\n"
	 "  -v               Be more verbose\n"
	 "  -h               Display this help\n"
	 "  -V               Display version information\n",
	 program_name);
}

static const std::string file2type(const char *filename)
{
  CPlayers::const_iterator	i;
  CSilentopl			opl;
  CPlayer			*p;

  for(i = CAdPlug::players.begin(); i != CAdPlug::players.end(); i++)
    if((p = (*i)->factory(&opl)))
      if(p->load(filename)) {
	delete p;
	return (*i)->filetype;
      } else
	delete p;

  message(MSG_WARN, "unknown filetype -- %s", filename);
  return UNKNOWN_FILETYPE;
}

static void db_add(const char *filename)
{
  CAdPlugDatabase::CRecord *record = CAdPlugDatabase::CRecord::factory(cfg.rtype);

  if(!record) {
    message(MSG_ERROR, "internal error (not enough memory?)");
    exit(EXIT_FAILURE);
  }

  record->key = file2key(filename);
  record->filetype = file2type(filename);
  if(record->filetype == UNKNOWN_FILETYPE) { delete record; return; }
  if(!record->user_read(std::cin, std::cout)) {
    message(MSG_ERROR, "data entry error");
    exit(EXIT_FAILURE);
  }

  if(mydb.lookup(record->key)) {
    message(MSG_NOTE, "replacing previous record -- %s", filename);
    mydb.wipe();
  }

  if(mydb.insert(record)) {
    message(MSG_NOTE, "added record, file type \"%s\" -- %s",
	    record->filetype.c_str(), filename);
  } else {
    delete record;
    message(MSG_ERROR, "error adding record to database -- %s", filename);
    exit(EXIT_FAILURE);
  }
}

static bool db_resolve(const char *filename)
/* Resolves and lists one entry from the database */
{
  if(mydb.lookup(file2key(filename))) {
    mydb.get_record()->user_write(std::cout);
    return true;
  } else {
    message(MSG_WARN, "no entry in database -- %s", filename);
    return false;
  }
}

static void db_remove(const char *filename)
/* Removes one entry from the database */
{
  if(mydb.lookup(file2key(filename))) {
    mydb.wipe();
    message(MSG_NOTE, "deleted entry -- %s", filename);
  } else
    message(MSG_WARN, "no entry in database, could not delete -- %s", filename);
}

static void copyright()
/* Print copyright notice and version information */
{
  printf("AdPlug database maintenance utility %s\n", CAdPlug::get_version().c_str());
  printf("Copyright (c) 2002 Riven the Mage <riven@ok.ru>\n"
	 "Copyright (c) 2002 Simon Peter <dn.tlp@gmx.net>\n");
}

static void db_error(bool dbokay)
/* Checks if database is open. Exits program otherwise */
{
  if(!dbokay) {	// Database could not be opened
    message(MSG_ERROR, "database could not be opened -- %s", cfg.db_file);
    exit(EXIT_FAILURE);
  }
}

/***** Main program *****/

int main(int argc, char *argv[])
{
  char		opt;
  bool		dbokay;
  unsigned int	i;

  // Extract program name from argv[0]
  program_name = strrchr(argv[0], '/') ? strrchr(argv[0], '/') + 1 : argv[0];

  // Parse options
  while((opt = getopt(argc, argv, "d:t:qvhV")) != -1)
    switch(opt) {
    case 'd': cfg.db_file = optarg; break;		// Set database file
    case 't': // Different record type
      for(i = 0; rtypes[i].typestr; i++)
	if(!strcmp(rtypes[i].typestr, optarg)) {
	  cfg.rtype = rtypes[i].type;
	  break;
	}

      if(!rtypes[i].typestr) {
	message(MSG_ERROR, "unknown record type -- %s", optarg);
	exit(EXIT_FAILURE);
      }
      break;
    case 'q': if(cfg.message_level) cfg.message_level--; break;	// Be more quiet
    case 'v': cfg.message_level++; break;	       	// Be more verbose
    case 'h': usage(); exit(EXIT_SUCCESS); break;	// Display help
    case 'V': copyright(); exit(EXIT_SUCCESS); break;	// Display version
    case '?': exit(EXIT_FAILURE);
    }

  // Load database file
  dbokay = mydb.load(cfg.db_file);

  // Check for commands
  if(argc == optind) {
    fprintf(stderr, "%s: need a command\n", program_name);
    fprintf(stderr, "Try '%s -h' for more information.\n", program_name);
    exit(EXIT_FAILURE);
  }

  // Parse commands
  if(!strcmp(argv[optind], "add")) {	// Add file to database
    if(++optind < argc) {
      for(;optind < argc; optind++)
	db_add(argv[optind]);
      mydb.save(cfg.db_file);
    } else {
      message(MSG_ERROR, "add -- missing file argument");
      exit(EXIT_FAILURE);
    }
  } else
  if(!strcmp(argv[optind], "list")) {	// List (files from) database
    db_error(dbokay);
    if(++optind < argc) {
      for(;optind < argc; optind++)
	db_resolve(argv[optind]);
    } else {
      mydb.goto_begin();
      do {
	mydb.get_record()->user_write(std::cout);
	printf("\n");
      } while(mydb.go_forward());
    }
  } else
  if(!strcmp(argv[optind], "remove")) {	// Remove files from database
    db_error(dbokay);
    if(++optind < argc) {
      for(;optind < argc; optind++)
	db_remove(argv[optind]);
      mydb.save(cfg.db_file);
    } else {
      message(MSG_ERROR, "remove -- missing file argument");
      exit(EXIT_FAILURE);
    }
  } else {
    message(MSG_ERROR, "unknown command -- %s", argv[optind]);
    exit(EXIT_FAILURE);
  }
}
